File: doctor_appinment.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type Doctor struct {
	Name         string
	Speciality   string
	Availability []string
	Rating       int
}

type Patient struct {
	Name               string
	BookedAppointments map[string]map[string]string
}

type AppointmentSystem struct {
	Doctors   []*Doctor
	Patients  []*Patient
	Bookings  map[string]map[string]string
	BookingID int
}

func NewAppointmentSystem() *AppointmentSystem {
	return &AppointmentSystem{
		Doctors:   []*Doctor{},
		Patients:  []*Patient{},
		Bookings:  map[string]map[string]string{},
		BookingID: 1000,
	}
}

func NewDoctor(name, speciality string, rating int) *Doctor {
	return &Doctor{
		Name:         name,
		Speciality:   speciality,
		Availability: []string{},
		Rating:       rating,
	}
}

func NewPatient(name string) *Patient {
	return &Patient{
		Name:               name,
		BookedAppointments: map[string]map[string]string{},
	}
}

func (d *Doctor) DeclareAvailability(slots []string) {
	d.Availability = append(d.Availability, slots...)
}

func (p *Patient) BookAppointment(bookingID, slot, doctorName string) {
	if p.BookedAppointments == nil {
		p.BookedAppointments = map[string]map[string]string{}
	}
	p.BookedAppointments[bookingID] = map[string]string{"slot": slot, "doctor": doctorName}
}

func (p *Patient) CancelAppointment(bookingID string) bool {
	if _, exists := p.BookedAppointments[bookingID]; exists {
		delete(p.BookedAppointments, bookingID)
		return true
	}
	return false
}

func (as *AppointmentSystem) RegisterDoctor(name, speciality string, rating int) {
	doctor := NewDoctor(name, speciality, rating)
	as.Doctors = append(as.Doctors, doctor)
	fmt.Printf("Welcome Dr. %s !!\n", name)
}

func (as *AppointmentSystem) MarkDocAvail(doctorName string, slots []string) {
	doctor := as.FindDoctorByName(doctorName)
	if doctor == nil {
		fmt.Println("Doctor not found!")
		return
	}

	for _, slot := range slots {
		if !as.IsValidSlot(slot) {
			fmt.Printf("Invalid slot format for Dr. %s. Slots must be in format 'hh:mm-hh:mm'\n", doctorName)
			return
		}

		startEnd := strings.Split(slot, "-")
		if len(startEnd) != 2 {
			fmt.Printf("Invalid slot format for Dr. %s. Slots must be in format 'hh:mm-hh:mm'\n", doctorName)
			return
		}
		doctor.DeclareAvailability(slots)
	}
	fmt.Println("Done Doc!")
}

func (as *AppointmentSystem) RankDoctorsByRating(speciality string) map[string]int {
	rankedDoctors := map[string]int{}
	for _, doctor := range as.Doctors {
		if doctor.Speciality == speciality {
			rankedDoctors[doctor.Name] = doctor.Rating
		}
	}
	return rankedDoctors
}

func (as *AppointmentSystem) ShowAvailBySpeciality(speciality string, rankingStrategy string) {
	fmt.Println(speciality)
	if rankingStrategy == "rating" {
		rankedDoctors := as.RankDoctorsByRating(speciality)
		for doctorName, _ := range rankedDoctors {
			as.ShowAvailForDoctor(doctorName)
		}
	} else {
		for _, doctor := range as.Doctors {
			if doctor.Speciality == speciality {
				as.ShowAvailForDoctor(doctor.Name)
			}
		}
	}
}

func (as *AppointmentSystem) ShowAvailForDoctor(doctorName string) {
	doctor := as.FindDoctorByName(doctorName)
	if doctor != nil {
		for _, slot := range doctor.Availability {
			fmt.Printf("Dr. %s: (%s)\n", doctor.Name, slot)
		}
		fmt.Println()
	}
}

func (as *AppointmentSystem) RegisterPatient(name string) {
	patient := NewPatient(name)
	as.Patients = append(as.Patients, patient)
	fmt.Printf("%s registered successfully.\n", name)
}

func (as *AppointmentSystem) BookAppointment(patientName, doctorName, slot string, waitlist bool) {
	patient := as.FindPatientByName(patientName)
	doctor := as.FindDoctorByName(doctorName)

	if patient == nil || doctor == nil {
		fmt.Println("Patient or Doctor not found!")
		return
	}

	if !waitlist && as.IsSlotBooked(slot) {
		fmt.Println("Slot already booked!")
		return
	}

	if !as.IsSlotAvailable(doctor, slot) {
		fmt.Println("Slot not available for booking!")
		return
	}

	bookingID := strconv.Itoa(as.BookingID + 1)
	as.BookingID++
	as.Bookings[bookingID] = map[string]string{"patient": patientName, "doctor": doctorName, "slot": slot}
	patient.BookAppointment(bookingID, slot, doctorName)
	doctor.Availability = RemoveString(doctor.Availability, slot)
	fmt.Printf("Booked. Booking id: %s\n", bookingID)
}

func (as *AppointmentSystem) CancelBookingID(bookingID string) {
	if _, exists := as.Bookings[bookingID]; !exists {
		fmt.Println("Booking not found!")
		return
	}

	slot := as.Bookings[bookingID]["slot"]
	doctorName := as.Bookings[bookingID]["doctor"]
	delete(as.Bookings, bookingID)

	for _, patient := range as.Patients {
		if patient.CancelAppointment(bookingID) {
			break
		}
	}

	as.FindDoctorByName(doctorName).Availability = append(as.FindDoctorByName(doctorName).Availability, slot)
	fmt.Println("Booking Cancelled")
}

func (as *AppointmentSystem) ShowAppointmentsBooked(patientName string) {
	patient := as.FindPatientByName(patientName)
	if patient == nil {
		fmt.Println("Patient not found!")
		return
	}

	for bookingID, slotInfo := range patient.BookedAppointments {
		fmt.Printf("Booking id: %s, Dr %s %s\n", bookingID, slotInfo["doctor"], slotInfo["slot"])
	}
}

func (as *AppointmentSystem) FindDoctorByName(name string) *Doctor {
	for _, doctor := range as.Doctors {
		if doctor.Name == name {
			return doctor
		}
	}
	return nil
}

func (as *AppointmentSystem) FindPatientByName(name string) *Patient {
	for _, patient := range as.Patients {
		if patient.Name == name {
			return patient
		}
	}
	return nil
}

func (as *AppointmentSystem) IsValidSlot(slot string) bool {
	return len(slot) == 11 && slot[2] == ':' && slot[5] == '-' && slot[8] == ':'
}

func (as *AppointmentSystem) IsSlotBooked(slot string) bool {
	for _, booking := range as.Bookings {
		if booking["slot"] == slot {
			return true
		}
	}
	return false
}

func (as *AppointmentSystem) IsSlotAvailable(doctor *Doctor, slot string) bool {
	for _, availableSlot := range doctor.Availability {
		if availableSlot == slot {
			return true
		}
	}
	return false
}

func RemoveString(slice []string, s string) []string {
	var result []string
	for _, str := range slice {
		if str != s {
			result = append(result, str)
		}
	}
	return result
}

func main() {
	system := NewAppointmentSystem()
	fmt.Println("Enter command (registerDoc, markDocAvail, showAvailByspeciality, registerPatient, bookAppointment, cancelBookingId, showAppointmentsBooked, exit):")
	var command string
	for {
		fmt.Scanln(&command)
		if command == "exit" {
			break
		}

		args := strings.Split(command, " ")
		action := args[0]

		switch action {
		case "registerDoc":
			name, speciality := args[1], args[2]
			rating := 4
			system.RegisterDoctor(name, speciality, rating)
		case "markDocAvail":
			name := args[1]
			slots := args[2:]
			system.MarkDocAvail(name, slots)
		case "showAvailByspeciality":
			speciality := args[1]
			rankingStrategy := "start_time"
			if len(args) > 2 {
				rankingStrategy = args[2]
			}
			system.ShowAvailBySpeciality(speciality, rankingStrategy)
		case "registerPatient":
			name := args[1]
			system.RegisterPatient(name)
		case "bookAppointment":
			patientName, doctorName, slot := args[1], args[2], args[3]
			waitlist := false
			if len(args) > 4 && args[4] == "true" {
				waitlist = true
			}
			system.BookAppointment(patientName, doctorName, slot, waitlist)
		case "cancelBookingId":
			bookingID := args[1]
			system.CancelBookingID(bookingID)
		case "showAppointmentsBooked":
			patientName := args[1]
			system.ShowAppointmentsBooked(patientName)
		default:
			fmt.Println("Invalid command!")
		}
		fmt.Println("Enter command:")
	}
}

File: doctor_appinment.php
<?php

class Doctor
{
    public $name;
    public $speciality;
    public $availability = [];
    public $rating;

    public function __construct($name, $speciality, $rating)
    {
        $this->name = $name;
        $this->speciality = $speciality;
        $this->rating = $rating;
    }

    public function declareAvailability($slots)
    {
        foreach ($slots as $slot) {
            $this->availability[] = $slot;
        }
    }
}

class Patient
{
    public $name;
    public $bookedAppointments = [];

    public function __construct($name)
    {
        $this->name = $name;
    }

    public function bookAppointment($bookingId, $slot, $doctorName)
    {
        $this->bookedAppointments[$bookingId] = ["slot" => $slot, "doctor" => $doctorName];
    }

    public function cancelAppointment($bookingId)
    {
        if (isset($this->bookedAppointments[$bookingId])) {
            unset($this->bookedAppointments[$bookingId]);
            return true;
        }
        return false;
    }
}

class AppointmentSystem
{
    public $doctors = [];
    public $patients = [];
    public $bookings = [];
    public $bookingId = 1000;

    public function registerDoctor($name, $speciality, $rating = 4)
    {
        $doctor = new Doctor($name, $speciality, $rating);
        $this->doctors[] = $doctor;
        echo "Welcome Dr. $name !!\n";
    }

    public function markDocAvail($doctorName, $slots)
    {
        $doctor = $this->findDoctorByName($doctorName);
        if (!$doctor) {
            echo "Doctor not found!\n";
            return;
        }
        foreach ($slots as $slot) {
            $slot = rtrim($slot, ',');

            if (!$this->isValidSlot($slot)) {
                echo "Invalid slot format for Dr. $doctorName. Slots must be in format 'hh:mm-hh:mm'\n";
                return;
            }

            list($start, $end) = explode('-', $slot);
            $startTime = strtotime($start);
            $endTime = strtotime($end);
            if (($endTime - $startTime) !== 3600) {
                echo "Invalid slot duration for Dr. $doctorName. Slots must be exactly 60 minutes long.\n";
                return;
            }
            $this->addSlotToDoctor($doctor, $slot);
        }

        echo "Done Doc!\n";
    }

    private function isValidSlot($slot)
    {
        $pattern = '/^(\d{1,2}|\d{2}):(\d{1,2}|\d{2})-(\d{1,2}|\d{2}):(\d{1,2}|\d{2})$/';
        return preg_match($pattern, $slot);
    }

    private function addSlotToDoctor($doctor, $slot)
    {
        list($startTime, $endTime) = explode('-', $slot);
        $doctor->availability[] = $slot;
    }

    public function rankDoctorsByRating($speciality)
    {
        $rankedDoctors = [];
        foreach ($this->doctors as $doctor) {
            if ($doctor->speciality === $speciality) {
                $rankedDoctors[$doctor->name] = $doctor->rating;
            }
        }
        arsort($rankedDoctors);
        return $rankedDoctors;
    }
    // Display slots in a ranked fashion
    public function showAvailByspeciality($speciality, $rankingStrategy = 'start_time')
    {
        echo $speciality;
        if ($rankingStrategy === 'rating') {
            $rankedDoctors = $this->rankDoctorsByRating($speciality);
            foreach ($rankedDoctors as $doctorName => $rating) {
                $this->showAvailForDoctor($doctorName, $speciality);
            }
        } else {
            foreach ($this->doctors as $doctor) {
                if ($doctor->speciality === $speciality) {
                    $this->showAvailForDoctor($doctor->name, $speciality);
                }
            }
        }
    }

    private function showAvailForDoctor($doctorName, $speciality)
    {
        $doctor = $this->findDoctorByName($doctorName);
        foreach ($doctor->availability as $slot) {
            echo "Dr. $doctor->name: ";
            echo "($slot) \n";
        }
        echo "\n";
    }

    public function registerPatient($name)
    {
        $patient = new Patient($name);
        $this->patients[] = $patient;
        echo "$name registered successfully.\n";
    }

    public function bookAppointment($patientName, $doctorName, $slot, $waitlist = false)
    {
        $patient = $this->findPatientByName($patientName);
        $doctor = $this->findDoctorByName($doctorName);

        if (!$patient || !$doctor) {
            echo "Patient or Doctor not found!\n";
            return;
        }

        if (!in_array($slot, $doctor->availability)) {
            echo "Slot not available for booking!\n";
            return;
        }

        if (!$waitlist && $this->isSlotBooked($slot)) {
            echo "Slot already booked!\n";
            return;
        }

        $bookingId = ++$this->bookingId;
        $this->bookings[$bookingId] = ["patient" => $patientName, "doctor" => $doctorName, "slot" => $slot];
        $patient->bookAppointment($bookingId, $slot, $doctorName);
        $this->removeSlotFromDoctor($doctor, $slot);
        echo "Booked. Booking id: $bookingId\n";
    }

    private function removeSlotFromDoctor($doctor, $slotToRemove)
    {
        $updatedAvailability = [];
        foreach ($doctor->availability as $slot) {
            if ($slot !== $slotToRemove) {
                $updatedAvailability[] = $slot;
            }
        }
        $doctor->availability = $updatedAvailability;
    }

    public function cancelBookingId($bookingId)
    {
        if (!isset($this->bookings[$bookingId])) {
            echo "Booking not found!\n";
            return;
        }

        $slot = $this->bookings[$bookingId]["slot"];
        $doctorName = $this->bookings[$bookingId]["doctor"];
        unset($this->bookings[$bookingId]);
        foreach ($this->patients as $patient) {
            if ($patient->cancelAppointment($bookingId)) {
                break; // Cancel only once for the given booking ID
            }
        }

        // Add cancelled slot back to doctor's availability
        $this->addSlotToDoctor($this->findDoctorByName($doctorName), $slot);
        echo "Booking Cancelled\n";
    }

    public function showAppointmentsBooked($patientName)
    {
        $patient = $this->findPatientByName($patientName);
        if (!$patient) {
            echo "Patient not found!\n";
            return;
        }

        foreach ($patient->bookedAppointments as $slot) {
            $bookingId = array_search($slot, array_column($this->bookings, 'slot'));
            $doctor = $this->bookings[$bookingId]["doctor"];
            echo "Booking id: $bookingId, Dr $doctor $slot\n";
        }
    }

    private function findDoctorByName($name)
    {
        foreach ($this->doctors as $doctor) {
            if ($doctor->name === $name) {

                return $doctor;
            }
        }
        return null;
    }

    private function findPatientByName($name)
    {
        foreach ($this->patients as $patient) {
            if ($patient->name === $name) {
                return $patient;
            }
        }
        return null;
    }

    private function isSlotBooked($slot)
    {
        foreach ($this->bookings as $booking) {
            if ($booking["slot"] === $slot) {
                return true;
            }
        }
        return false;
    }
}

$system = new AppointmentSystem();
echo "Enter command (registerDoc, markDocAvail, showAvailByspeciality, registerPatient, bookAppointment, cancelBookingId, showAppointmentsBooked, exit): \n";
while (true) {
    $command = readline();
    if ($command === 'exit') {
        break;
    }

    $args = explode(" ", $command);
    $action = $args[0];

    switch ($action) {
        case 'registerDoc':
            $name = $args[1];
            $speciality = $args[2];
            $system->registerDoctor($name, $speciality);
            break;
        case 'markDocAvail':
            $name = $args[1];
            $slots = array_slice($args, 2);
            $system->markDocAvail($name, $slots);
            break;
        case 'showAvailByspeciality':
            $speciality = $args[1];
            $system->showAvailByspeciality($speciality);
            break;
        case 'registerPatient':
            $name = $args[1];
            $system->registerPatient($name);
            break;
        case 'bookAppointment':
            $patientName = $args[1];
            $doctorName = $args[2];
            $slot = $args[3];
            $waitlist = isset($args[4]) && $args[4] === 'true';
            $system->bookAppointment($patientName, $doctorName, $slot, $waitlist);
            break;
        case 'cancelBookingId':
            $bookingId = $args[1];
            $system->cancelBookingId($bookingId);
            break;
        case 'showAppointmentsBooked':
            $patientName = $args[1];
            $system->showAppointmentsBooked($patientName);
            break;
        default:
            echo "Invalid command!\n";
            break;
    }
}
